### PR TITLE
[3143] Modifications to the JSON-RPC API for `execute` request

### LINKED ISSUE #3143: Overhaul graph traversal/execution strategy
- [x] #3168
- [x] Design work towards JSON-RPC new functionality
- [x] Create a front-end ticket when this is done, so we can update the pyk library

### PR DESCRIPTION
**Pull request by jberthold**
To accommodate the requirements on the future `execute` implementation, the API description was modified in the following way:

In the response to an `execute` request:
* added current `state` field
* added `rule` field (optional, filled for `cut-point-rule` or `terminal-rule`)
* renamed `states` to `next-states` (optional, filled for `branching` or `cut-point-rule`)
* avoid repetition of the `depth` field

See #3143 for additional information.

Fixes #3143

### Scope:
- [X] Requirements analysis and implementation design for RPC `execute`
- [X] Adaptation of RPC server stub code
- [x] Prototype implementation of the new `execute` back-end

### Estimate:

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

File: kore/src/Kore/JsonRpc.hs [hunk 1]
```diff
@@ -143,7 +141,10 @@ data HaltReason
 
 data ExecuteResult = ExecuteResult
     { reason :: HaltReason
-    , states :: [ExecuteState]
+    , depth :: !Depth
+    , state :: ExecuteState
+    , nextStates :: Maybe [ExecuteState]
```
Comment (goodlyrottenapple) [inline kore/src/Kore/JsonRpc.hs:146]: Is there a difference in meaning when `nextState` is `Nothing` vs `Just []` or is this purely so that the `ToJSON` instance does not include the `nextState` when `Nothing`?

File: kore/src/Kore/Exec/GraphTraversal.hs [hunk 2]
```diff
@@ -48,8 +48,8 @@ data TransitionResult a
     | -- | final state (e.g., goal state reached, side conditions hold)
       Final a
     | -- | not stuck, but also not final (maximum depth reached before
-      -- finishing the proof). Provides the "next" states for the result.
-      Stopped [a]
+      -- finishing the proof). Provides current and "next" states for the result.
+      Stop a [a]
```
Comment (goodlyrottenapple) [inline kore/src/Kore/Exec/GraphTraversal.hs:52]: Just curious, doesnt matter either way, but why the change?

Comment (goodlyrottenapple) [reply kore/src/Kore/Exec/GraphTraversal.hs:52]: (i.e. Stopped -> Stop)

File: kore/src/Kore/JsonRpc.hs (hunk 1 above)
Comment (jberthold) [reply kore/src/Kore/JsonRpc.hs:146]: Just for JSON. We can of course make it non-optional but it is clearly only needed for some of the reasons.

File: kore/src/Kore/Exec/GraphTraversal.hs (hunk 2 above)
Comment (jberthold) [reply kore/src/Kore/Exec/GraphTraversal.hs:52]: I thought `Stop/Stopped` in analogy to `Stuck _ :: TransitionResult` => `GotStuck _ _ :: TraversalResult`. 
`Stopped` is a good name for the `TraversalResult`, and we cannot have two `Stopped` constructors :man_shrugging:

File: kore/src/Kore/JsonRpc.hs (hunk 1 above)
Comment (goodlyrottenapple) [reply kore/src/Kore/JsonRpc.hs:146]: No I think that's absolutely fine. Just wanted to check. Thanks

File: kore/src/Kore/Exec/GraphTraversal.hs (hunk 2 above)
Comment (goodlyrottenapple) [reply kore/src/Kore/Exec/GraphTraversal.hs:52]: Fair enough

Comment (goodlyrottenapple) [reply kore/src/Kore/Exec/GraphTraversal.hs:52]: LGTM btw

File: docs/2022-07-18-JSON-RPC-Server-API.md [hunk 3]
```diff
@@ -87,52 +87,106 @@ If the verification of the `state` pattern fails, the following error is returne
...
-        "state": {"format":"KORE", "version":1, "term":{}},
-        "depth": 1
-      }
-    ],
-    "reason": "final-state"
+    "state": {
+      "state": {"format":"KORE", "version":1, "term":{}},
+      "predicate": {"format":"KORE", "version":1, "term":{}},
+      "substitution": {"format":"KORE", "version":1, "term":{}},
+    },
```
Comment (jberthold) [inline docs/2022-07-18-JSON-RPC-Server-API.md:108]: There is more work to do here... a `Substitution` is not equivalent to a `KorePattern` (rather a collection of pairs of `KorePattern`, or a map from `Text` to `KorePattern`)

File: kore/src/Kore/JsonRpc.hs [hunk 4]
```diff
@@ -120,9 +130,9 @@ instance FromRequest (API 'Req) where
     parseParams _ = Nothing
 
 data ExecuteState = ExecuteState
-    { state :: !KoreJson
-    , depth :: !Depth
-    , condition :: !(Maybe Condition)
+    { state :: KoreJson
+    , substitution :: Maybe KoreJson
```
Comment (jberthold) [inline kore/src/Kore/JsonRpc.hs:134]: See prior comment, `substitution`'s type has to change

File: kore/src/Kore/JsonRpc.hs [hunk 5]
```diff
@@ -298,6 +362,12 @@ respond
       where
         couldNotVerify err = ErrorObj "Could not verify KORE pattern" (-32002) err
 
+        serverError detail payload =
```
Comment (goodlyrottenapple) [inline kore/src/Kore/JsonRpc.hs:365]: Should the errors all have different codes? Per the JSON RPC spec, you should use -32000 to -32099 for user errors. Perhaps make all errors a big enum and then use [`fromEnum`](https://hackage.haskell.org/package/base/docs/Prelude.html#v:fromEnum) to assign the code?

File: docs/2022-07-18-JSON-RPC-Server-API.md (hunk 3 above)
Comment (ana-pantilie) [inline docs/2022-07-18-JSON-RPC-Server-API.md:108]: In keeping with the current terminology, I'd expect this to be something like:
```suggestion
    "state": {
      "term": {"format":"KORE", "version":1, "value":{}},
      "predicate": {"format":"KORE", "version":1, "value":{}},
      "substitution": {"format":"KORE", "version":1, "value":{}},
    },
```

File: docs/2022-07-18-JSON-RPC-Server-API.md [hunk 6]
```diff
@@ -87,52 +87,106 @@ If the verification of the `state` pattern fails, the following error is returne
...
   "result": {
-    "states": [
-      {
-        "state": {"format":"KORE", "version":1, "term":{}},
-        "depth": 1
-      }
-    ],
-    "reason": "final-state"
+    "state": {
+      "state": {"format":"KORE", "version":1, "term":{}},
```
Comment (ana-pantilie) [inline docs/2022-07-18-JSON-RPC-Server-API.md:105]:
```suggestion
      "term": {"format":"KORE", "version":1, "term":{}},
```

File: kore/src/Kore/JsonRpc.hs (hunk 4 above)
Comment (ana-pantilie) [reply kore/src/Kore/JsonRpc.hs:134]: Not necessarily, we can represent substitutions in K like: (X #Equals a) #And (Y #Equals b) ...

File: kore/src/Kore/JsonRpc.hs (hunk 5 above)
Comment (jberthold) [reply kore/src/Kore/JsonRpc.hs:364]: Good idea, but more cumbersome to keep the doc.s in sync.
For now I have arbitrarily used `-32032` (leaving some room for more user errors before this server error), and added to the documentation.

File: kore/src/Kore/Syntax/Json.hs [hunk 7]
```diff
@@ -94,11 +97,40 @@ prettyJsonOpts =
...
+'[t1 / X][t2 / Y]'becomes '#And ( X #Equals t1, #And ( Y #Equals t2, ... ))'.
+The result sort is fixed to a sort variable.
+-}
+fromSubstitution :: Substitution VariableName -> Maybe KoreJson
+fromSubstitution subst
+    | Substitution.null subst = Nothing
+    | otherwise =
+        Just
+            . fromTermLike
+            . foldl1 TermLike.mkAnd
```
Comment (goodlyrottenapple) [inline kore/src/Kore/Syntax/Json.hs:119]: this will technically produce

```
[t1 / x1][t2 / x2]..[tn / xn] ~> #And ( ... And (x1 #Equals t1, x2 #Equals t2) ..., xn #Equals tn )
```
... i think... never remember which way folds go...

**Review by goodlyrottenapple — APPROVED** [review]
LGTM

File: kore/src/Kore/Syntax/Json.hs (hunk 7 above)
Comment (jberthold) [reply kore/src/Kore/Syntax/Json.hs:119]: You are right, the code does not match the description. I'll fix the description then.

Comment (goodlyrottenapple) [reply kore/src/Kore/Syntax/Json.hs:125]: thanks :)
